keps/sig-multicluster/1645-multi-cluster-services-api/README.md

```diff
@@ -413,9 +415,19 @@ type ServiceExport struct {
         // +optional
         metav1.ObjectMeta `json:"metadata,omitempty"`
         // +optional
+        Spec ServiceExportSpec `json:"spec,omitempty"`
+        // +optional
         Status ServiceExportStatus `json:"status,omitempty"`
 }
 
+// ServiceExportSpec describes an exported service and extra exported information
+type ServiceExportSpec struct {
+        // +optional
+        ExportedLabels map[string]string `json:"exportedLabels"`
+        // +optional
+        ExportedAnnotations map[string]string `json:"exportedAnnotations"`
```

Comment:
You mean like that we would have some filters instead of the actual values? If so yes that might be one alternative approach that would work too, let's see what other think!

Reply:
If the ServiceExport stores the values, we need to describe how updates to those values are handled. I don’t think we’d ever want different values on a Service and its ServiceExport, so having only the names of the labels/annotations is sufficient IMO and avoids some complexity.

Reply:
Hmm maybe it's not clear but what I wrote in the KEP is not supposed to look at the corresponding Service at all, like all the labels/annotations exported would be in the ServiceExport spec a bit like what's done in gateway API (which makes me realize that I haven't posted the link to gateway API doc in the description: https://gateway-api.sigs.k8s.io/geps/gep-1762/#labels-and-annotations)

Reply:
Although I was originally in favor of not adding any fields to the ServiceExport object at all, but with @mikemorris suggestion in the meeting I was a bit inclined to do it a bit like GatewayAPI and tried to advocate for the reasons we would like to do that in this PR although a different approach is definitely doable!

Reply:
Ah, right — I interpreted the mention of Service in the section title, “Exporting labels/annotations from the Service/ServiceExport objects”, incorrectly! And in fact you do exclude this interpretation further down, “Implementations must not export the annotations or
labels from the metadata of the `Service` or `ServiceExport` resources.”

In that case it would be interesting to consider whether exported metadata is likely to be identical to the existing ServiceExport metadata, or have different values — storing values separately ensures that we can have different values in the ServiceExport and (generated) ServiceImport metadata, whereas a filter approach prevents that entirely.

Reply:
If we go on the filtering route there might be annotations that only make sense from a clustermesh POV. This is very Cilium specific but for instance we do have an annotation to prefer the local cluster or the remote cluster (this https://docs.cilium.io/en/latest/network/clustermesh/affinity/#enabling-global-service-affinity). So while it would work on the local (not derived) Service it would makes it a bit weird.

Also you might want to possibly have a different set of labels for your ServiceImport (and derived Service) from your local service. For instance, if you want to select only the derived Service and not potentially the local Service it might be a bit problematic to take the labels (even filtered) from the local Service. We can totally solve that by saying something along those lines in the KEP "if the implementation use a derived Service you must add the label `multicluster.kubernetes.io/derived-service: "true"` " which would be sufficient IMO.

No big blockers though just throwing additional constraints/requirements that we may consider...

Reply:
> We can totally solve that by saying something along those lines in the KEP "if the implementation use a derived Service you must add the label `multicluster.kubernetes.io/derived-service: "true"` which would be sufficient IMO.

We have a [similar requirement](https://gateway-api.sigs.k8s.io/geps/gep-1762/#automated-deployments) over in Gateway API for controllers to add a `gateway.networking.k8s.io/gateway-name: my-gateway` label to any generated resources, which we then depend on for testing whether any configured infrastructure labels/annotations have been set on generated resources - I would suggest naming this slightly more generic like `multicluster.kubernetes.io/derived-resource` but I think the concept is sound. I'm not sure if an explicit link to _which_ ServiceImport/Export it's derived from is necessary or if that could be inferred from the name, but it may be helpful.

Reply:
I would advise against the copy-filter pattern, we tried that on a Gateway API implementation I worked on previously, and it was [a bit burdensome and a source of confusion for users](https://github.com/hashicorp/consul-api-gateway/issues/601).

Reply:
The semantics depends on the use case we're trying to solve. My interpretation was that the intent was to transfer verbatim certain labels/annotations values from the exported `Service` to the derived `ServiceImport` (and derived `Service` if the implementation creates one), in which case we'd want only the names specified. But if the intent is to populate derived labels/annotations solely from the `ServiceExport`, then, of course, we'd need the values specified as well. With that approach, if the user wants the derived resource to reflect the exported `Service`'s values, then they would have to know that info when creating the `ServiceExport` and also update the `ServiceExport` accordingly on changes, if necessary. A bit more of a burden as @skitt noted.

I'm not clear on the exact use case for Cilium (Submariner doesn't have a use case for this). @MrFreezeex re: "_you might want to possibly have a different set of labels for your ServiceImport (and derived Service) from your local service_", is this part of Cilium's use case or is this just hypothetical?

Reply:
> I'm not clear on the exact use case for Cilium (Submariner doesn't have a use case for this). @MrFreezeex re: "you might want to possibly have a different set of labels for your ServiceImport (and derived Service) from your local service", is this part of Cilium's use case or is this just hypothetical?

For labels the use case would be beneficial for many implementations I think. This is mostly for third party controllers that have selectors. Regarding ServiceImport specifically I don't think there is any use case so far (but there might be in the future!) but for the derived Service if you can add some labels any controllers that have a service selector would be able to make use of this. For instance, in the ServiceMonitor resource of Prometheus operator you have a service selector which means that if you start populating labels to your derived Service you can theoretically (if you sync EndpointSlice) do some multicluster scraping. If you didn't have labels this would be much harder to integrate with those type of controllers...

About annotations, we have many annotations in Cilium that change the behavior of a Service which may or may not be related directly to the clustermesh implementation (for instance there's an annotation to load balance multiple HTTP2 stream to different backends instead of load balancing at the connection levels like you would typically do).

IMO the best choice for labels and annotations sync would be either:
- Take labels and annotations it from the metadata Service directly. I am not entirely convinced right now about a label/annotations filters per object so I would probably favor only an annotation prefix filter at the controller level (that can probably be configured when installing the software implementing MCS-API, but the KEP should stay vague about that anyway).
- Like the current version this PR is describing: have the labels/annotations in the spec of ServiceExport that are exported verbatim without the local Service involved. This allows to fully decouple labels and annotations of the final ServiceImport (and then the derived Service for most implementation) which is a good thing for labels (I think?) but for annotation could be both nice and weird. The nice part is that you can put specific clustermesh annotation there (which removes the awkwardness of adding to a regular Service which in isolation has nothing to do with clustermesh/multi cluster). The bad part is that you can have annotations that modify the behavior of the Service entirely (such as the topology aware annotation if we stay in the standard annotations) which would end up only in the ServiceImport and its derived Service while not present on the local Service which is kind of arguable because it's not like any of the other fields...

There is also a middle ground/tweak to the first option, which is to merge the annotations (and potentially labels) of both the ServiceExport and the local Service before actually "exporting" it but I am worried in term of UX/coherence here as it's not like any other fields and not obvious...

Reply:
Argh actually I was checking about labels and for instance ArgoCD use a label to track the Application (https://argo-cd.readthedocs.io/en/latest/user-guide/annotations-and-labels/) so unlike what I just said above there is definitely a need for labels filtering too (at least to make sure that they don't conflict)... So what you were suggesting @tpantelis @skitt looks like a solid option too.

Reply:
There’s probably use cases for both approaches. This PR describes copying from the ServiceExport which is fine. The filter approach could achieved via key prefixes, either well known or configured, or specific keys. Perhaps this should be left up to implementations to define and implement as they see fit and either leave it out of the KEP or perhaps mention it as optional and not define specifics.

Reply:
> Perhaps this should be left up to implementations to define and implement as they see fit and either leave it out of the KEP or perhaps mention it as optional (in the Alternatives section) but not define specifics.

WDYM? Because like if we add some fields to the ServiceExport implementation must at least make some use of those fields (whether it's the field to filter like you are proposing or directly the map like the current version of this PR).

Are you suggesting that the new fields would do something (TBD what exactly) and outside of that the implementation may have some well known prefix somehow built-in that he could take from the local Service directly for instance? Meaning that in the KEP we would describe what those fields do without enforcing that it is the only way that an implementation may use to export labels and annotations?

Reply:
So we've been discussing two potential ways to propagate labels/annotations to derived resource(s) with potential use cases for each.

1) Specify the exact keys and values to copy in the `ServiceExport` spec.

2) Specify the keys that are copied from the `Service` instance. 

Both approaches have pros and cons depending on the use case. So should we pick one or the other to formally define/describe in the KEP and possibly discuss the other in the _Alternatives_ section? Or should we define both to give users ultimate flexibility on how and from where to propagate labels/annotations?

Whichever is formally defined, will it be required or optional for implementations to support? This seems geared towards supporting a derived `Service`, which is not required by the spec. Submariner does  not create derived `Services` and would have no use to propagate labels/annotations to the `ServiceImport`.

Given that, should we then not formally define any way to propagate labels/annotations and let implementations handle it (or not) as they see fit? Both approaches could be implemented via key prefixes without adding constructs to the `ServiceExport` CRD.

Reply:
> I think we could make it optional though

yes I think it should be optional given that the motivation is to support derived `Services` which are implementation-specific and not required.

Reply:
I pushed some new changes to clearly mark labels and annotation export as optional and detailed more about the filtering alternative, let me know what do you think about this new version :pray:.

Reply:
> * Specify the exact keys and values to copy in the `ServiceExport` spec.
> 
> * Specify the keys that are copied from the `Service` instance.

How about a CEL expression that takes a key and returns true or false? It's expensive to evaluate but the set of keys is typically stable so the filter shouldn't be recomputed often.

Potentially two expressions: one for labels, one for annotations

Why is this good?
- flexible
- lets you do things like wildcards but with an exception
- very configurable handling for unrecognized labels / annotations
- allows for rules like "export either both of these labels or neither of them" or "export label A or B, but never both"
- extensible (I think) to filtering on values as well

Reply:
AIUI server-side apply lets us go further and decide whether to become a field manager for a particular key, and if so to:
- take over management
- jointly manage the key
- set / update the value but only if there is no other declared manager
- skip this key and not copy it to the target

A CEL expression could return which of those four options to pick, per label / annotation key.

Reply:
> AIUI server-side apply lets us go further and decide whether to become a field manager for a particular key

I don't think this is a case where server-side apply would be relevant here because this is not a conflict problem within the same cluster between different actors.

The context is that we need to figure out a set of labels/annotations that are getting exported from each "local Services" in a way that would reasonably not result in a "cross-cluster conflict"  even if third party tools are used that might add extra labels/annotations (kubectl with the last applied annotation, argocd with the application labels are example of that).

> How about a CEL expression that takes a key and returns true or false? It's expensive to evaluate but the set of keys is typically stable so the filter shouldn't be recomputed often.

This is quite an interesting idea to apply CEL here. I see it has a "copy-filter" approach but in a much more flexible way. But it's also way more complex I am not sure if as a user of an MCS-API implementation, I would want to write some CEL expressions to filter my labels/annotations and also probably making sure that this expression would be the same on all similar ServiceExport across clusters...

So IMO this has the same benefits of "copy-filter" approach (for instance a list of prefix) which is consistency with the rest of the API: the source of truth is on the Service itself with more flexibility but also more complex for users.

I am not entirely sure if this is really better than the the way this KEP is currently proposing which is trying to skip this copy-filter problem entirely by adding dedicated fields in the spec that won't be touched by third party tooling which may result in a conflict when reconciling those cross clusters.

Reply:
@MrFreezeex I'd like the CEL expression idea to appear in the list of alternatives; it doesn't have to get listed as the winning choice.

What I disagreed with, lightly, was the implication that we had to choose between exactly 2 design options here. Actually there are more.

Reply:
Well we never considered CEL but AFAICT it's still is fundamentally a "copy-filter" approach which we discussed here and with some points against it in `Exporting labels/annotations from the Service/ServiceExport objects`. The only thing that might not be relevant for the CEL proposal is the last point about flexibility which is traded by more complexity. Will try to rephrase at to that last point to include writing CEL expressions.

Reply:
(OK, this is resolved)